src/java/test/org/apache/zookeeper/server/quorum/QuorumPeerMainTest.java

```diff
@@ -888,4 +888,127 @@ public void testWithOnlyMinSessionTimeout() throws Exception {
+            mt[i] = new MainThread(i, clientPorts[i], quorumCfgSection);
+            mt[i].start();
+            zk[i] = new ZooKeeper("127.0.0.1:" + clientPorts[i], ClientBase.CONNECTION_TIMEOUT, this);
+        }
+
+        waitForAll(zk, States.CONNECTED);
+
+        // we need to shutdown and start back up to make sure that the create session isn't the first transaction since
+        // that is rather innocuous.
+        for (int i = 0; i < SERVER_COUNT; i++) {
+            mt[i].shutdown();
+        }
+
+        waitForAll(zk, States.CONNECTING);
+
+        for (int i = 0; i < SERVER_COUNT; i++) {
+            mt[i].start();
+            // Recreate a client session since the previous session was not persisted.
+            zk[i] = new ZooKeeper("127.0.0.1:" + clientPorts[i], ClientBase.CONNECTION_TIMEOUT, this);
+        }
+
+        waitForAll(zk, States.CONNECTED);
+
+        // 2. kill all followers
+        int leader = -1;
+        Map<Long, Proposal> outstanding = null;
+        for (int i = 0; i < SERVER_COUNT; i++) {
+            if (mt[i].main.quorumPeer.leader != null) {
+                leader = i;
+                outstanding = mt[leader].main.quorumPeer.leader.outstandingProposals;
+                // increase the tick time to delay the leader going to looking
+                mt[leader].main.quorumPeer.tickTime = 10000;
+            }
+        }
+        LOG.warn("LEADER {}", leader);
+
+        for (int i = 0; i < SERVER_COUNT; i++) {
+            if (i != leader) {
+                mt[i].shutdown();
+            }
+        }
+
+        // 3. start up the followers to form a new quorum
+        for (int i = 0; i < SERVER_COUNT; i++) {
+            if (i != leader) {
+                mt[i].start();
+            }
+        }
+
+        // 4. wait one of the follower to be the leader
+        for (int i = 0; i < SERVER_COUNT; i++) {
+            if (i != leader) {
+                // Recreate a client session since the previous session was not persisted.
+                zk[i] = new ZooKeeper("127.0.0.1:" + clientPorts[i], ClientBase.CONNECTION_TIMEOUT, this);
+                waitForOne(zk[i], States.CONNECTED);
+            }
+        }
+
+        // 5. send a create request to leader and make sure it's synced to disk,
+        //    which means it acked from itself
+        try {
+            zk[leader].create("/zk" + leader, "zk".getBytes(), Ids.OPEN_ACL_UNSAFE,
+                CreateMode.PERSISTENT);
+            Assert.fail("create /zk" + leader + " should have failed");
+        } catch (KeeperException e) {
+        }
+
+        // just make sure that we actually did get it in process at the
+        // leader
+        Assert.assertTrue(outstanding.size() == 1);
+        Proposal p = (Proposal) outstanding.values().iterator().next();
+        Assert.assertTrue(p.request.getHdr().getType() == OpCode.create);
+
+        // make sure it has a chance to write it to disk
+        Thread.sleep(1000);
```

Comment:
There is a lot of `Thread.sleep()` going on and I would like to find a way to minimize that. Apache infra can occasionally be quite slow (it can starve threads) and tests with many `Thread.sleep()`s in them have historically been quite flaky.

So, to the extent that it is possible. I would like to minimize occurrences of `Thread.sleep()`, or at least those outside the context of retry logic.

So perhaps, we can throw `p.qvAcksetPairs.get(0).getAckset().contains(leader);` in a loop waiting one second between iterations.

w.r.t. step 6, we can wait for the leader to enter the looking state.

What do you think?

Reply:
I will see if I can make it work.  I agree I would love to kill as many of the sleeps as possible.

Reply:
I was able to do what you said and drop the 1 second sleep, but the sleep at step 6 I am going to need something else because the leader is only in the looking state for 2 ms.  Leader election happens way too fast for us to be able to catch that by polling.  

If I remove the 4 second sleep it does not trigger the error case, I don't completely know why.  I'll spend some time looking at it, but if you have any suggestions I would appreciate it.

Reply:
@revans2 take a look at `testElectionFraud`, specifically: https://github.com/apache/zookeeper/blob/master/src/java/test/org/apache/zookeeper/server/quorum/QuorumPeerMainTest.java#L383 and https://github.com/apache/zookeeper/blob/master/src/java/test/org/apache/zookeeper/server/quorum/QuorumPeerMainTest.java#L397

You can manually start and stop leader election, I think that may solve this problem.